### PR TITLE
Reimplement data structure and add GC

### PR DESCRIPTION
**Pull request by kitten**
`6.47kB` up to `7.01kB` minzipped (gotta do some house keeping soon I guess, but let's be honest: this is worth it! :tada:)

tl;dr This adds garbage collecting (with automatic reference counting) and is pretty exciting, however it does sound like it’d use more memory and be slower, but as it turns out it’s faster ✨ (About ~25%; more measurements needed)

## What?

This PR intends to add a new data structure for our store’s in memory storage. It also then adds gar age collection.

## How?

For full garbage collection we need to be able to delete full entities (all their fields or “records”). Our desired behaviour is that when all “links” (references from one entity to another) disappear and the entity has no links referring to it anymore, we delete the entity.

This wasn’t possible before because we stored all fields of an entity, which we call “records” separately. We do have “connections” which remember all fields for a given entity that have arguments, but this doesn’t store all fields. To keep things efficient we now create a dictionary object for each entity. So our map of records and links is now nested. (Entity key then field key). This keeps things nice and efficient.

We also don’t intend to add garbage collection using “tracing”. Traced garbage collection would scan all links then remove all entities that don’t have any links referring to it anymore. This is a slow and indeterministic process.

Instead we use reference counting (“RC”). We keep a counter of how many links are referring to an entity (by its entity key). When that number reaches zero, we queue the entity key for deletion.

Deletion is scheduled to happen one tick after an operation, so we don’t perform GC too often. During GC, we go through all entities and delete them (also taking their links into account, so it’s an unrolled recursion)

If an entity is still needed for optimistic data, we create a “GC lock”. This is a list of entities per optimistic update that are still needed. When an optimistic update is deleted (“cleared”) we then delete the lock.

## Is this ARC?

Yes, this is "automatic reference counting" rather than "mark & sweep garbage collection". We count the references to each entity and delete the ones during a GC run that have reached zero references.

If you're familiar with ARC then you may know that this means that circular references become problematic. When two entities would reference each other (or a longer chain becomes circular) then these entities wouldn't be garbage collected.

We see this as a healthy trade-off. We're dealing with in-memory data here and chances are, if you've fetched this data you're ok with using up your memory to hold on to it. Where this really becomes an issue is offline storage. I believe once we address that we'll add a second mark & sweep algorithm that kicks in on startup, so we can clean up the offline storage. Until then though this is out of scope.

## Breaking Changes

All apps should continue to work as they did before. But due to the data structure changes this also changes some methods on the `Cache` interface (the store).

### `resolveConnections` was deleted

Instead you can now use `inspectFields` which replaces it and is already used now on `relayPagination` and `simplePagination`.

This accepts a partial entity or an entity key and returns an array of all fields as a `FieldInfo` type. This contains the `fieldKey`, the `fieldName`, and its `arguments` (object or null).

```ts
  inspectFields(entity: Data | string | null): FieldInfo[];
```

### `resolveValueOrLink` was deleted

Instead you can now use `resolveFieldByKey` which is more similar to `cache.resolve`.

It accepts a partial entity or an entity key, and a `fieldKey` (`fieldName` + `args`). It then returns the field's value or link.

```ts
  resolveFieldByKey(entity: Data | string | null, fieldKey: string): DataField;
```

### `keyOfField` was added

This allows you to create a `fieldKey` manually if you need to use `resolveFieldByKey` (or something similar) instead of `resolve`.

```ts
  keyOfField(
    fieldName: string,
    args?: Variables | null | undefined
  ): string | null;
```

### CODE REVIEW

File: src/operations/query.ts [hunk 1]
```diff
@@ -522,7 +516,7 @@ const resolveResolverResult = (
   } else if (result === null || result === undefined) {
```
Comment (JoviDeCroock) [inline src/operations/query.ts]: I think this is causing the bug with `simplePagination`
